### PR TITLE
Upsert by specifying a column list

### PR DESCRIPTION
**Pull request by ilyaminati**

### CODE REVIEW

File: R/upsert2.R [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  }
+  numRows = nrow(dat)
+
+  if (numRows == 0) {
+    warning("Empty dataset provided, will not try uploading")
+    return(FALSE)
+  }
+
+  if (getOption("verbose")) {
+    message("The provided data.frame has ", numRows, " rows")
```
Comment (sicarul) [inline R/upsert2.R]: I agree with your use of message instead of print for this function, but why ask for the verbose option? If the user wants to suppress the messages, they can use `supressMessages` when calling the function

File: R/upsert2.R [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+#' host='my-redshift-url.amazon.com', port='5439',
+#' user='myuser', password='mypassword',sslmode='require')
+#'
+#' rs_cols_upsert_table(df=nx, dbcon=con, table_name='testTable',
+#' bucket="my-bucket", split_files=4, values=c('b'), keys=c('a'))
+#'}
+#' @seealso \url{https://docs.aws.amazon.com/redshift/latest/dg/merge-specify-a-column-list.html}
+#' @seealso \url{http://docs.aws.amazon.com/redshift/latest/dg/tutorial-configuring-workload-management.html}
+#' @export
+rs_cols_upsert_table = function(dat,
```
Comment (sicarul) [inline R/upsert2.R:40]: Why change from df to dat? Just wondering why it is different here

File: R/upsert2.R [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    )
+    if (getOption("verbose")) {
+      message(qu)
+    }
+    res <- queryStmt(dbcon, qu)
+    if (getOption("verbose")) {
+      message(res, " rows")
+    }
+
+    qu <- sprintf('DELETE FROM %s \nUSING %s \nWHERE %s',
```
Comment (sicarul) [inline R/upsert2.R:117]: I might be misunderstanding something, but why are you deleting after updating?

Comment (sicarul) [reply R/upsert2.R:117]: OHHH You are deleting on the staging table, looking at the existing table, genius, never mind me 👍

File: DESCRIPTION [hunk 4]
```diff
@@ -1,7 +1,7 @@
 Package: redshiftTools
 Type: Package
 Title: Redshift Tools
-Version: 0.3.900
+Version: 0.3.1000
```
Comment (sicarul) [inline DESCRIPTION]: Can you revert this change? .900 is to follow hadley's advise on versioning:
http://r-pkgs.had.co.nz/description.html

Though i see that i mistakenly put 900 instead of 9000 😓 (Or instead of reverting, go to 9000)

**Comment by ilyaminati** [comment]
@sicarul Happy to follow up discussion here in the PR, but didn't get the comments you said that you sent.

**Review by sicarul — CHANGES_REQUESTED** [review]
This is the review, sorry i forgot to submit it before.

File: R/upsert2.R (hunk 1 above)
Comment (sicarul) [reply R/upsert2.R]: I only made this comment once but applies to all usage of the verbose option.

Comment (ilyaminati) [reply R/upsert2.R]: Either `supressMessages` or `getOption("verbose")` seems fine to me. Your call, happy to take out the `if`. Probably best to make it consistent across the entire package, although I'm not sure I'm up for updating the other `rs_*` functions right now.

File: R/upsert2.R (hunk 2 above)
Comment (ilyaminati) [reply R/upsert2.R:40]: `df` is an existing base function, `?df`.

File: DESCRIPTION (hunk 4 above)
Comment (ilyaminati) [reply DESCRIPTION]: Sure, I'll go to 9000.

File: R/upsert2.R (hunk 2 above)
Comment (sicarul) [reply R/upsert2.R:40]: Ok, makes sense, i will edit the other functions later

File: DESCRIPTION (hunk 4 above)
Comment (ilyaminati) [reply DESCRIPTION]: Looking closely, Hadley's advice is that 9000 is the _fourth_ component of a version number, e.g., `0.0.0.9000`. Not sure it makes sense for the third component.

File: R/upsert2.R (hunk 1 above)
Comment (sicarul) [reply R/upsert2.R]: I'd prefer to make them appear by default, and allow them to be suppressed, because a novice user may have no idea what is going on, also i think it's a good default to debug in logs for batch processes, in case things go wrong

**Comment by ilyaminati** [comment]
@sicarul I think I've resolved all your concerns. Ready to commit?

**Comment by sicarul** [comment]
Yes, thanks for the contribution!

**Review by sicarul — APPROVED** [review]

**Comment by ilyaminati** [comment]
Wonderful, thanks for the quick review!
